feat: add components docs (#3)
kubesaw/kubesaw.github.io: docs/components.md

# Components
This page provides a brief description of each KubeSaw component.

## Host Operator

> The host-operator is _KubeSaw's control-plane_, is responsible for managing the users, spaces, notifications and monitoring the kubesaw instance.

Reviewer: ```suggestion
The host-operator is _KubeSaw's control-plane_, and is responsible for managing the users, spaces, notifications and monitoring the kubesaw instance.
```

Reviewer: or you can take rafaela suggestion and make it even in both operators

Author: yes let's go with that if you agree, since I've already applied that before seeing your comment :)

Author: pls see https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/fbf8a46fd9c56a396e1267059d20b72a34816728

> One of the resources managed by host operator is the UserSignup resource, which is the source of truth for all user accounts. All other user-related resources are created from the UserSignup.

Reviewer: ```suggestion
One of the resources managed by host operator is the UserSignup resource, which is the source of truth for all user accounts. All the other user-related resources are created from the UserSignup.
```

Author: thank you!

> At the moment, there can be only one host operator deployed per KubeSaw instance.

Reviewer 2: ```suggestion
At the moment, only one host operator can be deployed per KubeSaw instance.
```

Author: thanks!

> At the moment, only one host operator can be deployed per KubeSaw instance.

Reviewer 3: Why is this important enough to be part of the high level description? IMHO, this is more of a detail that we don't have to include here. Unless this is the only place in the docs where we want to talk about the host operator as such, of course.

Author: I agree, we should cover that in the section dedicated to host operator details. Removed in https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/52b1a277db24b9daed76bedc7a8e065ae08c4425 
thanks!

The source code is available here [host-operator](https://github.com/codeready-toolchain/host-operator)

## Member Operator

> The member operator is _KubeSaw's data-plane_, is responsible for provisioning and managing the user namespaces and all the configuration inside those namespaces.

Reviewer 2: ```suggestion
The member operator is _KubeSaw's data-plane_, which is responsible for provisioning and managing the user namespaces and all the configurations inside those namespaces.
```

Author: thanks!

> There can be multiple member operators deployed as part of the same KubeSaw instance, ideally there should be a member operator per each cluster dedicated to the user workload (member cluster).

Reviewer: ```suggestion
There can be multiple member operators deployed as part of the same KubeSaw instance, ideally there should be one member operator per  cluster, dedicated to the user workload (member cluster).
```

Author: thanks a lot!

> There can be multiple member operators deployed as part of the same KubeSaw instance, ideally there should be one member operator per each cluster, dedicated to the user workload (member cluster).

Reviewer 3: I find the second part of the sentence a little bit confusing. What about:

There can be multiple member operators deployed as part of the same KubeSaw instance. Member operators are usually deployed per-cluster but can even co-exist in the same cluster.

Author: Thanks for the suggestion, applied in https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/52b1a277db24b9daed76bedc7a8e065ae08c4425

The source code is available here [member-operator](https://github.com/codeready-toolchain/member-operator)

## Registration Service
The Registration Service is a standalone web application that runs on the same cluster as the host operator
(host cluster).
It provides a number of REST endpoints that the signup page can use to allow a user to sign up for a KubeSaw account. 

> The Registration Service application makes use of third party libraries such as Gin (an HTTP web framework)

Reviewer 3: I think this whole sentence is superfluous.

Author: I kind of agree, but I wanted to add something that could allow the reader to understand how the registration service is different from the other components ( it's a "traditional" web app and not a controller ) , but we can remove it from here if we think it's not worth mentioning this here.

and Kubernetes client libraries (for interacting with the Kubernetes API) among others.

### Proxy
The Proxy is part of the registration service application deployment,

> basically it's another set of REST endpoints implemented using the echo framework.

Reviewer 3: I think the "basically" part is not necessary here.

Author: removed in https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/52b1a277db24b9daed76bedc7a8e065ae08c4425

It can be used in a multi-cluster environment to forward requests without the need to specify the cluster context,
the proxy infers the target cluster from the user information and the target workspace/namespace of the request.

Both Proxy and Registration-service are optional components and not required for having a fully functional KubeSaw instance.
The registration service can be useful if your KubeSaw service requires a signup mechanism/frontend.

> The proxy can help, in the case of a multi-cluster solution, with removing some of the _cluster awareness_ and friction for the end users or clients interacting with the member clusters.

Reviewer 4: Another key feature of the proxy is that it accepts the SSO tokens, not the member cluster kube token. I think it's worth mentioning.

Author: Good point! 
Pls see https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/37314524892e3fbea3ef36e2e5c95340a67c2ed3 , and let me know if the wording makes sense or feel free to adjust it in case.

Thank you 🙏

The source code for both registration-service and proxy is available here [registration-service](https://github.com/codeready-toolchain/registration-service)

## KSCTL

> ksctl is a command-line tool that helps you with managing your instance of KubeSaw.

Reviewer 2: ```suggestion
ksctl is a command-line tool that helps you manage your KubeSaw instance.
```

Author: thank you 🙏

> ksctl is a command-line tool that helps you manage your KubeSaw instance.

Reviewer 3: I'd add something along the lines of:

It provides commands for managing the clusters, spaces and users in Kubesaw.

Author: Good point, done in https://github.com/kubesaw/kubesaw.github.io/pull/3/commits/52b1a277db24b9daed76bedc7a8e065ae08c4425 

Thanks!
